### PR TITLE
Add generic spksrc.supported.mk

### PR DESCRIPTION
**Pull request by th0ma7**
### TL;DR;
Rework, optimize and simplify generic `mk` calls involving a major refactory but most importantly major clean-up of the code in hope to simplify future maintenance:
- [x] remove tons of duplicated code by using `spksrc.common-rules.mk`.  This invovles `*clean`, `dependency-*`, `changelog` and a new `rustup` such as in getting the rustc info using `make rustup show` for instance.
- [x] allow using `make all-supported` and `make all-latest` from anywhere applicable (i.e. `cross`, `spk`  and `diyspk`)
- [x] add a framework self-testing facility (to be incorporated into github-action later to trigger on any changes to `mk/*`)
- [x] allow `DEPENDS` from `toolchain`

### Self-test facility calls (as of March 6th, more to be included):
- `make testall`
- `make test clean`
- `make test depend`
- `make test digests`
- `make test download`
- `make test toolchain`
- `make test test-dependency-*`

TODO:
- `make test rust`
- `make test cmake`
- `make test gnu-configure`
- `make test meson`
- `make test native`
- ...

Superseeds #5998

## Checklist

- [ ] Build rule `all-supported` completed successfully
- [ ] New installation of package completed successfully
- [ ] Package upgrade completed successfully (Manually install the package again)
- [ ] Package [functionality was tested](https://github.com/SynoCommunity/spksrc/wiki/Package-Update-Policy#tests-checks)
- [ ] Any needed [documentation](https://github.com/SynoCommunity/spksrc/wiki/Create-documentation) is updated/created

### Type of change

- [ ] Bug fix
- [ ] New Package
- [ ] Package update
- [x] Includes small framework changes
- [ ] This change requires a documentation update (e.g. Wiki)

### TODO
- [x] `cc`
- [x] `cmake`
- [x] `dotnet`
- [x] `go`
- [x] `install-resources`
- [x] `meson`
- [x] `rust`
- [ ] `spk`

#### Toolchain -> COMPLETE
- [x] `~/spksrc$ make -C toolchain/syno-x64-7.1`
- [x] `~/spksrc$ make -C toolchain/syno-x64-7.1 clean`
- [x] `~/spksrc/toolchain/syno-x64-7.1$ make`
- [x] `~/spksrc/toolchain/syno-x64-7.1$ make clean`
- [x] `~/spksrc/toolchain/syno-x64-7.1$ make changelog`
```
git log --pretty=format:"- %s" -- /home/spksrc/all-supported-fix/spksrc/toolchain/syno-x64-7.1
- synocli-devel: New package - debugging, target python wheel building (#5225)
- update generic x64 archs (#5347)
- Add toolchains for DSM 7.1 (#5320)
```

#### `native` -> COMPLETE
- [x] `~/spksrc$ make -C native/nasm`
- [x] `~/spksrc$ make -C native/nasm clean`
- [x] `~/spksrc/native/nasm$ make`
- [x] `~/spksrc/native/nasm$ make clean`
- [x] `~/spksrc/native/elixir$ make changelog`
```
git log --pretty=format:"- %s" -- /home/spksrc/all-supported-fix/spksrc/native/elixir
- erlang: update to 24.3 (#5814)
- use make_install_options (#5350)
- Update erlang and rabbitmq with openssl 1.1.1l (#4913)
- rabbitmq: update to v3.8.16 (#4648)
- Add RabbitMQ and dedicated Erlang package (#4437)
```
- [x] `~/spksrc/native/elixir$ make dependency-list`
```
elixir: native/elixir native/erlang 
```
- [x] `~/spksrc/native/elixir$ make dependency-flat`
```
native/elixir
native/erlang
```
- [x] `~/spksrc/native/elixir$ make dependency-tree`
```
+ elixir 1.14.5
\ + erlang 24.3
```
#### `cross` -> COMPLETE
`cc` -> test with `xz`
`rust` -> test with `bat`
`cmake` -> test with `lz4`
`cmake` + `ninja` -> test with `zstd`
`cmake` + `main-depends.mk` + unsupported -> test with `intel-gmmlib`
- [x] `~/spksrc/cross/<app>$ make arch-qoriq-6.2.4`
- [x] `~/spksrc$ make -C cross/<app> arch-qoriq-6.2.4`
- [x] `~/spksrc/cross/<app>$ make TCVERSION=6.2.4 ARCH=qoriq`
- [x] `~/spksrc$ make -C cross/<app> TCVERSION=6.2.4 ARCH=qoriq`
- [x] `~/spksrc/cross/<app>$ PARALLEL_MAKE=max make -j4 all-supported`
- [x] `~/spksrc$ PARALLEL_MAKE=max make -C cross/<app> -j4 all-supported`
`===>> confirmed native get built first`
`===>> confirmed status-build.log with system-time at end of build (conditional to ~/spksrc/make setup)`
- [x] `~/spksrc/cross/<app>$ make clean`
- [x] `~/spksrc$ make -C cross/bat clean`
- [x] `~/spksrc/cross/<app>$ make changelog`
- [x] `~/spksrc/cross/<app>$ make dependency-flat`
- [x] `~/spksrc/cross/<app>$ make dependency-list`
- [x] `~/spksrc/cross/<app>$ make dependency-tree`
- [x] `~/spksrc/cross/<app>$ make rustup show`
```
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/spksrc/all-supported-fix/spksrc/cross/bat/../../distrib/rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu (default)
stage2-powerpc-unknown-linux-gnuspe

installed targets for active toolchain
--------------------------------------

aarch64-unknown-linux-gnu
armv5te-unknown-linux-gnueabi
armv7-unknown-linux-gnueabi
armv7-unknown-linux-gnueabihf
i686-unknown-linux-gnu
powerpc-unknown-linux-gnu
x86_64-unknown-linux-gnu

active toolchain
----------------

stable-x86_64-unknown-linux-gnu (default)
rustc 1.75.0 (82e1608df 2023-12-21)
```

### CODE REVIEW

**Comment by th0ma7** [comment]
@mreid-tt and @hgy59 while trickier than it looks this code fixes the dependency evaluation duration (current 3m20s as expected).

I'll pursue this question to remove code duplication in makefiles and overall simplification where possible.  My intent is to include other targets, `spk.mk`, `cross-cmake.mk` and `cross-rust.mk` so `make all-supported` works out always.

**Comment by th0ma7** [comment]
@hgy59 any interest in updating dotnet as well?  As that would require more in-depth changes to it...

**Comment by th0ma7** [comment]
@hgy59 and @mreid-tt working on long overdue major simplifications to the `spk.mk` file.  Current woks locally and needs further testing, mainly kernel related, github-action and publishing (although preliminary local tests works ok).

Basically it does the following:
1. extracts all the kernel, build and publishing code from spk and redirect them into other files, with hopefully a better and simplified narrative for exit calls and dependencies.
2. remove duplicates of `clean` and `smart-clean` being in everywhere
3. enable tracing

As mentionned, there's still testing to be done, along with further optimizing the code.  But your initial thoughts on it would be welcomed to see if any gaps have been overlooked.

My hope from here is that future maintenance will be simplified thus allowing easier inclusion of other longer term goals, mainly related to python wheel building complexity for instance (while not related, overall code simplification and clean-up will ease integrating subsequent changes).

Thoughts on this welcomed.

**Comment by mreid-tt** [comment]
@th0ma7, big thanks for your awesome work on optimizing spksrc's core. Just a heads up, I don't have a local environment for testing. I rely solely on building and publishing from GitHub. But hey, I'm here to help out in any way I can with a little guidance from you. Even though some of the kernel-level stuff goes over my head, I'm still keen to chip in where possible.

**Comment by th0ma7** [comment]
You really first need an environment to play with.  Worst case if you get to install an Ubuntu using WSL on windows would be a start, from there following the lxc howto on our wiki.  I can assist over discord as needed.

**Comment by mreid-tt** [comment]
> You really first need an environment to play with. Worst case if you get to install an Ubuntu using WSL on windows would be a start, from there following the lxc howto on our wiki. I can assist over discord as needed.

Thank you for your support. I primarily use a Mac with Apple Silicon for my computing needs. Consequently, when I install Ubuntu via VMware, it's the ARM version. While this setup works well for spkrepo, I've discovered that spksrc necessitates an x86 platform, as indicated in the wiki. Hence, I currently lack a local environment for spksrc. If there's a workaround for ARM, I'd welcome further discussion on Discord.

**Comment by th0ma7** [comment]
Currently broken part of my last set of changes yesterday night, will let you know once resolved.

**Comment by th0ma7** [comment]
Now fixed.  Also started documenting every calls to confirm things are working as expected... 

If anyone has some idea on how to do this, having a github-action rule that "TEST" all the common calls when an `mk/*` gets modified would be really awesome.  Thus ensuring API stays always functional.

**Comment by Safihre** [comment]
@th0ma7 Could you provide a TLDR version of what this does? And what to test?

**Comment by th0ma7** [comment]
@Safihre certainly, here's a few highlights (to be added in top summary):
- [x] remove tons of duplicated code by using `spksrc.common-rules.mk`.  This invovles `*clean`, `dependency-*`, `changelog` and a new `rustup` such as in getting the rustc info using `make rustup show` for instance.
- [x] allow using `make all-supported` and `make all-latest` from anywhere applicable (i.e. `cross`, `spk`  and `diyspk`)
- [x] add a framework self-testing facility (to be incorporated into github-action later to trigger on any changes to `mk/*`)
- [x] allow `DEPENDS` from `toolchain`

So all in all, major refactory but most importantly major clean-up of the code in hope to simplify future maintenance.

**Comment by hgy59** [comment]
> * allow using `make all-supported` and `make all-latest` from anywhere applicable (i.e. `cross`, `spk`  and `diyspk`)

regarding the _`from anywhere`_  there is an issue in `spksrc.python.mk`.

working on homeassistant package I wanted to create dedicated wheels with (temp) diyspk Makefiles. When using pre-built python in `spk/python311` it did not work for packages in `diyspk` folder. 
This is caused by line 8 in spksrc.python.mk
```
PYTHON_PACKAGE_ROOT = $(realpath $(shell pwd)/../$(PYTHON_PACKAGE)/work-$(ARCH)-$(TCVERSION))
```
And can be fixed like this:
```
PYTHON_PACKAGE_ROOT = $(realpath $(shell pwd)/../../spk/$(PYTHON_PACKAGE)/work-$(ARCH)-$(TCVERSION))
```

In the context of this PR you should fix this too, as it could probably benefit of the new variable `$(BASEDIR)` defined in spksrc.common.mk.

---

Another cleanup proposal:
There are several statements `$(shell pwd)` but some mk files use $(PWD) (defined in `spksrc.directories.mk` as `PWD := $(shell pwd)`.
Find candidates by `grep pwd *.mk`.

**Comment by th0ma7** [comment]
@hgy59 now rebased against master - done.

**Comment by th0ma7** [comment]
> regarding the from anywhere there is an issue in spksrc.python.mk.

@hgy59 will definitively look at it.

> There are several statements $(shell pwd) but some mk files use $(PWD) (defined in spksrc.directories.mk as PWD := $(shell pwd).

Well, further reading, it should all be replaced with `$(CURDIR)` which in theory always work in makefiles... something else added to the pile.

Let me know if you find other things that doesn't work on this branch in comparison with master.  Considering the depth of the changes I'll take all the testing assistance possible.

**Comment by hgy59** [comment]
@th0ma7 one issue found:

`cross/openjpeg` does not build anymore.

this affects all ffmpeg and the imagemagick package.

```
*** No targets specified and no makefile found.  Stop.
```

it is using `spksrc.cross-cmake.mk` with `CMAKE_USE_NINJA = 1` but I have no idea...

**Comment by hgy59** [comment]
@th0ma7 another issue, not related to this PR

`spk/gateone` fails to build the wheels. This is a python2 only package and since we delcared `spk/python2` as `BROKEN`, we should mark `spk/gateone` as `BROKEN` too.

**Comment by th0ma7** [comment]
@hgy59 both meson + cmake using ninja are now fixed, thnx for identifying it.  Also fixed gateone while at it.

**Comment by th0ma7** [comment]
> There are several statements $(shell pwd) but some mk files use $(PWD) (defined in spksrc.directories.mk as PWD := $(shell pwd).

@hgy59 I've now migrated the framework to using makefile default `$(CURDIR)`.  Note that I haven't changed `cross/*` packages that make use of `$(PWD)` variable as this would trigger too many packages to be built.  As such for the moment I've kept :
```
# For legacy reasons keep $(PWD) call
PWD := $(CURDIR)
```

> regarding the from anywhere there is an issue in spksrc.python.mk.

Also now fixed.

**Comment by th0ma7** [comment]
Remaining: enhance the testing facility.

**Comment by th0ma7** [comment]
@hgy59 I've rebased against master and hit the following:
```
	providers/libdefault.a providers/libcommon.a  -ldl -pthread -latomic 
/home/spksrc/all-supported-fix/spksrc/toolchain/syno-88f6281-6.2.4/work/arm-marvell-linux-gnueabi/bin/../lib/gcc/arm-marvell-linux-gnueabi/4.6.4/../../../../arm-marvell-linux-gnueabi/bin/ld: cannot find -latomic
collect2: ld returned 1 exit status
make[4]: *** [Makefile:4768: libcrypto.so.3] Error 1
make[4]: *** Waiting for unfinished jobs....
make[4]: Leaving directory '/home/spksrc/all-supported-fix/spksrc/cross/openssl3/work-88f6281-6.2.4/openssl-3.1.5'
make[3]: *** [Makefile:2854: build_sw] Error 2
make[3]: Leaving directory '/home/spksrc/all-supported-fix/spksrc/cross/openssl3/work-88f6281-6.2.4/openssl-3.1.5'
make[2]: *** [../../mk/spksrc.compile.mk:44: compile_target] Error 2
make[2]: Leaving directory '/home/spksrc/all-supported-fix/spksrc/cross/openssl3'
make[1]: *** [../../mk/spksrc.supported.mk:75: build-arch-88f6281-6.2.4] Error 1
make[1]: Leaving directory '/home/spksrc/all-supported-fix/spksrc/cross/openssl3'
```
Most probably related to: https://github.com/SynoCommunity/spksrc/pull/6025

**Comment by hgy59** [comment]
> @hgy59 I've rebased against master and hit the following:
> 
> ```
> 	providers/libdefault.a providers/libcommon.a  -ldl -pthread -latomic 
> /home/spksrc/all-supported-fix/spksrc/toolchain/syno-88f6281-6.2.4/work/arm-marvell-linux-gnueabi/bin/../lib/gcc/arm-marvell-linux-gnueabi/4.6.4/../../../../arm-marvell-linux-gnueabi/bin/ld: cannot find -latomic
> collect2: ld returned 1 exit status
> make[4]: *** [Makefile:4768: libcrypto.so.3] Error 1
> make[4]: *** Waiting for unfinished jobs....
> make[4]: Leaving directory '/home/spksrc/all-supported-fix/spksrc/cross/openssl3/work-88f6281-6.2.4/openssl-3.1.5'
> make[3]: *** [Makefile:2854: build_sw] Error 2
> make[3]: Leaving directory '/home/spksrc/all-supported-fix/spksrc/cross/openssl3/work-88f6281-6.2.4/openssl-3.1.5'
> make[2]: *** [../../mk/spksrc.compile.mk:44: compile_target] Error 2
> make[2]: Leaving directory '/home/spksrc/all-supported-fix/spksrc/cross/openssl3'
> make[1]: *** [../../mk/spksrc.supported.mk:75: build-arch-88f6281-6.2.4] Error 1
> make[1]: Leaving directory '/home/spksrc/all-supported-fix/spksrc/cross/openssl3'
> ```
> 
> Most probably related to: #6025

@th0ma7 are you sure, that the patch `cross/openssl3/patches/88f6281/001-avoid-the-use-of-libatomic.patch` was applied?

**Comment by th0ma7** [comment]
@hgy59 good catch, something broke on the patch side, I'll investigate that.

**Comment by th0ma7** [comment]
@publicarray does this error ring any bell to you?  Relatively to ` No space left on device`, only for `x64`:
```
[Build (x64-6.2.4)](https://github.com/SynoCommunity/spksrc/actions/runs/8226855325/job/22494072123)
System.IO.IOException: No space left on device : '/home/runner/runners/2.314.1/_diag/Worker_20240311-024107-utc.log' at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset) at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite() at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder) at System.Diagnostics.TextWriterTraceListener.Flush() at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id) at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message) at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message) at GitHub.Runner.Worker.Worker.RunAsync(String pipeIn, String pipeOut) at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args) System.IO.IOException: No space left on device : '/home/runner/runners/2.314.1/_diag/Worker_20240311-024107-utc.log' at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset) at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite() at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder) at System.Diagnostics.TextWriterTraceListener.Flush() at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id) at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message) at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message) at GitHub.Runner.Common.Tracing.Error(Exception exception) at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args) Unhandled exception. System.IO.IOException: No space left on device : '/home/runner/runners/2.314.1/_diag/Worker_20240311-024107-utc.log' at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset) at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite() at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder) at System.Diagnostics.TextWriterTraceListener.Flush() at System.Diagnostics.TraceSource.Flush() at GitHub.Runner.Common.TraceManager.Dispose(Boolean disposing) at GitHub.Runner.Common.TraceManager.Dispose() at GitHub.Runner.Common.HostContext.Dispose(Boolean disposing) at GitHub.Runner.Common.HostContext.Dispose() at GitHub.Runner.Worker.Program.Main(String[] args)
[Build (x64-7.1)](https://github.com/SynoCommunity/spksrc/actions/runs/8226855325/job/22494072222)
System.IO.IOException: No space left on device : '/home/runner/runners/2.314.1/_diag/Worker_20240311-024108-utc.log' at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset) at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite() at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder) at System.Diagnostics.TextWriterTraceListener.Flush() at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id) at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message) at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message) at GitHub.Runner.Worker.Worker.RunAsync(String pipeIn, String pipeOut) at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args) System.IO.IOException: No space left on device : '/home/runner/runners/2.314.1/_diag/Worker_20240311-024108-utc.log' at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset) at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite() at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder) at System.Diagnostics.TextWriterTraceListener.Flush() at GitHub.Runner.Common.HostTraceListener.WriteHeader(String source, TraceEventType eventType, Int32 id) at GitHub.Runner.Common.HostTraceListener.TraceEvent(TraceEventCache eventCache, String source, TraceEventType eventType, Int32 id, String message) at System.Diagnostics.TraceSource.TraceEvent(TraceEventType eventType, Int32 id, String message) at GitHub.Runner.Common.Tracing.Error(Exception exception) at GitHub.Runner.Worker.Program.MainAsync(IHostContext context, String[] args) Unhandled exception. System.IO.IOException: No space left on device : '/home/runner/runners/2.314.1/_diag/Worker_20240311-024108-utc.log' at System.IO.RandomAccess.WriteAtOffset(SafeFileHandle handle, ReadOnlySpan`1 buffer, Int64 fileOffset) at System.IO.Strategies.BufferedFileStreamStrategy.FlushWrite() at System.IO.StreamWriter.Flush(Boolean flushStream, Boolean flushEncoder) at System.Diagnostics.TextWriterTraceListener.Flush() at System.Diagnostics.TraceSource.Flush() at GitHub.Runner.Common.TraceManager.Dispose(Boolean disposing) at GitHub.Runner.Common.TraceManager.Dispose() at GitHub.Runner.Common.HostContext.Dispose(Boolean disposing) at GitHub.Runner.Common.HostContext.Dispose() at GitHub.Runner.Worker.Program.Main(String[] args)
```

**Comment by hgy59** [comment]
> Relatively to ` No space left on device`, only for `x64`:

I don't think this is related to a specifc archs. The two x64 builds just are the latest completing and there are already 3.9 GB artifacts created.
When we started with the gibhub build action we had to clean each package after build to keep the used space small. 
Since we can't do a clean for python and ffmeg packages as those are reused by dependents, I suppose we reach the limit again (here with python311 ffmpeg5 and ffmpeg4).

I may be wrong with this guess and the space limit might be per runner. Then it could be x64 related. AFAICT the x64 packages are always larger than arm packages. And the ffmpeg packages for x64 are larger anyway, due to additional dependencies included.

**Comment by th0ma7** [comment]
@hgy59 what if implementing a `make clean-source` so all the source directories gets deleted, thus including all unnecessary build-time object files.  This could be ran after every package build on GitHub?

Actually, further thinking, this could be default behavior while keeping the `.package-<status>` files ...

Although if i recall this cannot really be tested and only works once tentative pr is merged for subsequent action builds?

**Comment by th0ma7** [comment]
heads-up @SynoCommunity/developers Last round of testing now with https://github.com/SynoCommunity/spksrc/pull/6038 implemented.  If all goes well I intent to merge in around a week or two when I'll have the needed cycles to repond to issues that may be found afterwards (because even though I tested it intensively, something will certainly break).

In the meantime I'll pursue work on the included testing framework to make it more efficient.

**Comment by th0ma7** [comment]
@hgy59 I'm now ready to merge this PR.  I cannot test publishing although code in theory did not change.  This will then allow me to pursue work on the python311 update (#6040).

Note that I expect that something, somehow, will not work as expected as this PR is quite invasive.  Although the current `make test all` which build 25GB+ with no error is a good indicative that problems, if any, should be relatively minor.

Any objections before I merge this?

**Comment by hgy59** [comment]
> @hgy59 I'm now ready to merge this PR. I cannot test publishing although code in theory did not change. This will then allow me to pursue work on the python311 update (#6040).
> 
> Note that I expect that something, somehow, will not work as expected as this PR is quite invasive. Although the current `make test all` which build 25GB+ with no error is a good indicative that problems, if any, should be relatively minor.
> 
> Any objections before I merge this?

@th0ma7 thanks for your huge and great work. 
The only issue i found so far, is that after a make native-clean (or clean in native/rust-qoriq) it does not build rust-qoriq when building a package depending on rust for qoriq.
It probably depends on the rust installation related to building the toolchain(s). I often face this issue because I cache the toolchains for all my different work copies but do not cache the native packages. And it was an issue when testing dedicated versions of rust in #6050. 
This is all related to rust and should be addressed in a dedicated PR.

BTW I didn't figure out what you really solved with _"allow `DEPENDS` from `toolchain`"_

**Review by hgy59 — APPROVED** [review]
LGTM

**Comment by th0ma7** [comment]
> thanks for your huge and great work. 

Thnx, now fingers crossed nothing breaks.

> The only issue i found so far, is that after a make native-clean (or clean in native/rust-qoriq) it does not build rust-qoriq when building a package depending on rust for qoriq. It probably depends on the rust installation related to building the toolchain(s). I often face this issue because I cache the toolchains for all my different work copies but do not cache the native packages. And it was an issue when testing dedicated versions of rust in #6050. This is all related to rust and should be addressed in a dedicated PR.

This is because when `rust-qoriq` is "installed" it uses `rustup toolchain link` which creates, well, a link between the `distrib/rustup/toolchains/stage2-powerpc-unknown-linux-gnuspe` and `native/rust-qoriq/work-native/install/usr/local`.  Therefore when cleaning-up native, it removes that symlink'ed folder.  Further, the `.rustc_done` status file is still showing as complete in the `toolchain/syno-qoriq-6.2.4/work` thus it will never re-try to install.  The way to fix this temporarely is to clean-up the `toolchain/syno-qoriq-6.2.4` folder and re-invoke make.  The more permanent way would be to force copying over the rust toolchain files into the `distrib/rustup` folder when building native (instead of the rustup link), so subsequent cleanup doesn't affect the installation.

> BTW I didn't figure out what you really solved with _"allow `DEPENDS` from `toolchain`"_

Instead of enforcing within the code a `make -C native/rust-qoriq` when `RUST_BUILD_TOOLCHAIN = 1`, I've added the `depend` functionality to the toolchain makefile so I now use the following (and much simpler) directly into `toolchain/syno-qoriq-6.2.4/Makefile`:
```
OPTIONAL_DEPENDS = native/rust-qoriq

ifneq ($(RUST_BUILD_TOOLCHAIN),1)
DEPENDS = native/rust-qoriq
endif
```
So with this train of thoughts, you could probably now delete the `.depend_done` and re-invoke make from syno-qoriq-6.2.4 directory to enforce re-installing the rust toolchain.

**Comment by th0ma7** [comment]
Just so we know as reference, invoking a full `make test all` takes a lot of disk space:
```
spksrc@spksrc:~$ du -sh all-supported-fix
83G	all-supported-fix
```

And while at it, @hgy59 thnx for the review.
